### PR TITLE
PLN-544: Add 128K context budget awareness to judges

### PR DESCRIPTION
**Pull request by mikeangstadt**
## Summary

- **128K context budget awareness for judges:** When `CLOSEDLOOP_CONTEXT_LIMIT` is set and indicates a context window smaller than 200K tokens, the context manager dynamically computes a reduced token budget and sets `context_128k_mode=true`. The `run-judges` skill reads this flag and skips all judges whose estimated prompt would exceed the available budget, recording them as `final_status=3` CaseScores with a fixed `"Skipped:"` justification.
- **Dynamic token budget in context-manager-for-judges:** Replaces the hardcoded 30K token ceiling with `min(30000, CLOSEDLOOP_CONTEXT_LIMIT - 98000)`, falling back to 30K when the env var is unset. Budget tables now use percentage-based formulas instead of fixed token values.
- **`CLOSEDLOOP_CONTEXT_LIMIT` forwarding in run-loop.sh:** The orchestration loop now exports `CLOSEDLOOP_CONTEXT_LIMIT` (when set by the caller) so downstream subagents and skills can detect the reduced context window.
- **Skipped-judge validation support:** `validate_judge_report.py` gains a `SKIP_SENTINEL` constant and `CaseScore.is_skipped()` method so that skipped judges with empty metrics still pass report validation.
- **Test hardening:** 180+ lines of new tests covering skipped-judge tolerance (top-level and metric-level justification, all-skipped reports, mixed reports). Also sanitizes `CLOSEDLOOP_COMMAND`/`LAST_CLAUDE_COMMAND` env vars in `test_run_loop_failure_marker.py` to prevent ambient shell state from flipping assertions.
- **Pyright config:** Adds `venvPath`/`venv` to `pyproject.toml` for correct virtual-env resolution.

## Changed Files

| File | Change |
|------|--------|
| `plugins/code/scripts/run-loop.sh` | Forward `CLOSEDLOOP_CONTEXT_LIMIT` to subagents |
| `plugins/code/tools/python/test_run_loop_failure_marker.py` | Sanitize env vars leaking into tests |
| `plugins/code/.claude-plugin/plugin.json` | Version bump → 1.12.0 |
| `plugins/judges/agents/context-manager-for-judges.md` | Dynamic token budget, 128K mode detection |
| `plugins/judges/skills/run-judges/SKILL.md` | Step 0.5 budget check, skip logic, error CaseScores |
| `plugins/judges/skills/run-judges/scripts/validate_judge_report.py` | `SKIP_SENTINEL`, `is_skipped()`, relaxed empty-metrics validation |
| `plugins/judges/skills/run-judges/scripts/test_validate_judge_report.py` | 180+ lines of skipped-judge tests |
| `plugins/judges/.claude-plugin/plugin.json` | Version bump → 1.8.3 |
| `pyproject.toml` | Add `venvPath`/`venv` for Pyright |

## Test Plan

- [x] All existing `pytest` tests pass
- [x] New `TestSkippedJudges` class covers `is_skipped()` for top-level and metric-level justification, non-skipped statuses, and edge cases
- [x] Skipped-judge reports (single, all-skipped, mixed) pass `validate_report()`
- [x] `test_run_loop_failure_marker.py` env sanitization prevents false positives from ambient shell state

---
Loop ID: 019e2c30-0f0f-7608-a6d1-a2b233433ade
Artifact: https://app.closedloop.ai/implementation-plans/PLN-544

### CODE REVIEW

File: plugins/judges/skills/run-judges/SKILL.md [hunk 1]
```diff
@@ -742,10 +866,37 @@ Each judge returns a **CaseScore** JSON object:
...
+      "metric_name": "{metric}_score",
+      "threshold": 0.8,
+      "score": 0.0,
+      "justification": "Skipped: artifact exceeded context budget after compression"
+    }
+  ]
+}
+```
+
+The justification string MUST be exactly `"Skipped: artifact exceeded context budget after compression"` — do not include token counts or other dynamic text. This fixed string allows downstream consumers to distinguish skipped judges from runtime failures.
```
Comment (thadeusb) [inline plugins/judges/skills/run-judges/SKILL.md:952]: This says justification MUST be exactly the fixed string, no dynamic text. But line ~1331 in the recovery section says "Justification must state the skip reason explicitly: 128K mode is active, estimated tokens, and available budget." An LLM executing this skill will get two contradictory instructions. Pick one.

File: plugins/judges/skills/run-judges/SKILL.md [hunk 2]
```diff
@@ -538,12 +538,130 @@ The run-judges skill supports three artifact types with different judge configur
...
+      "metric_name": "{primary_metric_name}_score",
+      "threshold": 0.8,
+      "score": 0.0,
+      "justification": "Skipped: artifact exceeded context budget after compression"
+    }
+  ]
+}
+```
+
+Substitute `{judge-name}` with the actual agent identifier (e.g., `dry-judge`, `ssot-judge`), `{primary_metric_name}` with the judge's primary metric name (e.g., `dry_score`, `ssot_score`), and the token counts with the computed values.
```
Comment (thadeusb) [inline plugins/judges/skills/run-judges/SKILL.md]: Sub-step D says "Substitute {judge-name} with the actual agent identifier... and the token counts with the computed values." But the template has no token count placeholders. The justification string is hardcoded with no {token} fields. This reads like a stale edit that forgot to remove the token-count substitution instruction after switching to a fixed string.

File: plugins/judges/skills/run-judges/scripts/test_validate_judge_report.py [hunk 3]
```diff
@@ -908,3 +944,181 @@ def test_prd_report_fails_under_feature_category(self, tmp_path: Path) -> None:
...
+        }
+
+        report_path = tmp_path / "plan-judges.json"
+        report_path.write_text(json.dumps(report, indent=2))
+
+        valid, message = validate_report(report_path, category="plan")
+        assert valid is True, f"Expected valid report with skipped judge via metric, got: {message}"
+
+    def test_multiple_skipped_judges_pass_validation(self, tmp_path: Path) -> None:
+        """Report with multiple skipped judges (all have case_ids) passes validation."""
```
Comment (thadeusb) [inline plugins/judges/skills/run-judges/scripts/test_validate_judge_report.py:1232]: The SKILL.md says "either all judges fit or none fit for a given run." So the real 128K scenario is all 16 skipped, zero normal. This test only skips 3 of 16. Add a test where every judge in the report is a skipped CaseScore with empty metrics.

File: plugins/judges/.claude-plugin/plugin.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "judges",
   "description": "LLM Judges plugin",
-  "version": "1.6.1",
+  "version": "1.8.3",
```
Comment (thadeusb) [inline plugins/judges/.claude-plugin/plugin.json]: Main is at 1.7.1, this PR sets 1.8.3 from a base of 1.6.1. Needs a rebase. Also 1.8.3 is a PATCH, but per CLAUDE.md this is a feature addition (new skip logic in run-judges), so it should be a MINOR bump like 1.8.0.

File: plugins/code/.claude-plugin/plugin.json [hunk 5]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "code",
   "description": "Code and planning framework plugin",
-  "version": "1.11.16",
+  "version": "1.12.0",
```
Comment (thadeusb) [inline plugins/code/.claude-plugin/plugin.json]: Main is at 1.11.20. This sets 1.12.0 from a stale base of 1.11.16. This will either merge-conflict or silently downgrade the version. Needs a rebase.

File: plugins/judges/skills/run-judges/SKILL.md (hunk 1 above)
Comment (mikeangstadt) [reply plugins/judges/skills/run-judges/SKILL.md:952]: Fixed — the recovery section (line 1415) now references the fixed justification string from the Path 2 template: `"Skipped: artifact exceeded context budget after compression"`. No dynamic text.

File: plugins/judges/skills/run-judges/SKILL.md (hunk 2 above)
Comment (mikeangstadt) [reply plugins/judges/skills/run-judges/SKILL.md]: Fixed — removed the stale "token counts with the computed values" phrase from sub-step D. The description now says the justification string is fixed with no dynamic text, consistent with the template.

File: plugins/judges/skills/run-judges/scripts/test_validate_judge_report.py (hunk 3 above)
Comment (mikeangstadt) [reply plugins/judges/skills/run-judges/scripts/test_validate_judge_report.py:1232]: Fixed — added `test_all_judges_skipped_passes_validation` which skips all 16 plan judges and verifies the report passes validation. All 83 tests pass.

File: plugins/judges/.claude-plugin/plugin.json (hunk 4 above)
Comment (mikeangstadt) [reply plugins/judges/.claude-plugin/plugin.json]: Fixed — rebased onto main. Judges version is now 1.8.0 (MINOR bump for feature addition, per CLAUDE.md). Resolved merge conflicts in SKILL.md (kept both error_reason docs from main and skip CaseScore docs from this PR) and test file (kept both test classes).

File: plugins/code/.claude-plugin/plugin.json (hunk 5 above)
Comment (mikeangstadt) [reply plugins/code/.claude-plugin/plugin.json]: Fixed — rebased onto main (1.11.20). Code plugin version is now 1.12.0 (MINOR bump). No merge conflicts in code plugin files.

File: plugins/judges/skills/run-judges/SKILL.md [hunk 6]
```diff
@@ -562,12 +562,130 @@ The run-judges skill supports three artifact types with different judge configur
...
+When 128K mode is active, estimate the token footprint for a typical judge prompt:
+
+```bash
+# Estimate preamble token count (common_input_preamble.md + artifact_preamble.md)
+PREAMBLE_CHARS=$(wc -c < "$COMMON_PREAMBLE_PATH")
+ARTIFACT_PREAMBLE_CHARS=$(wc -c < "$ARTIFACT_PREAMBLE_PATH")
+TOTAL_PREAMBLE_CHARS=$(( PREAMBLE_CHARS + ARTIFACT_PREAMBLE_CHARS ))
+
+# Estimate judge-input.json token count
+JUDGE_INPUT_CHARS=$(wc -c < "$CLOSEDLOOP_WORKDIR/judge-input.json")
```
Comment (shafty023) [inline plugins/judges/skills/run-judges/SKILL.md]: The budget check counts only `$CLOSEDLOOP_WORKDIR/judge-input.json`, but `judge-input.json` is just the descriptor envelope with relative paths. The judges then load the mapped source-of-truth artifacts, so a large PRD/Feature file or context bundle can pass this gate and still overflow the 128K context. Please compute the estimate from the actual mapped artifact files plus preambles/output reserve, and add a regression where the envelope is small but a referenced artifact is too large.

File: plugins/judges/skills/run-judges/SKILL.md [hunk 7]
```diff
@@ -562,12 +562,130 @@ The run-judges skill supports three artifact types with different judge configur
...
+**When to run:** After context preparation (Step 0 / Prerequisites Check) completes and before launching any judge batch.
+
+**Purpose:** In 128K context mode the available context window is significantly smaller than the standard 200K window. If a judge's estimated prompt (preamble + `judge-input.json` content) would exceed the remaining available context, launching that judge is likely to fail or produce a truncated/unusable result. Judges that cannot fit must be skipped and recorded as error CaseScores so the overall workflow still produces a complete report.
+
+#### Sub-step A: Detect 128K Mode
+
+**For plan and code artifact types** (context-manager produces a context file):
+
+```bash
+CONTEXT_FILE="$CLOSEDLOOP_WORKDIR/plan-context.json"   # or code-context.json
```
Comment (shafty023) [inline plugins/judges/skills/run-judges/SKILL.md]: For code mode this needs to read the canonical `.closedloop-ai/context/code-context.json` first. As written, the budget check points at root `code-context.json`, so a normal code run can miss `metadata.context_128k_mode`, default the guard off, and launch code judges without the new overflow protection. Please mirror the mapper path order and keep root `code-context.json` only as the legacy fallback.

File: plugins/judges/skills/run-judges/scripts/validate_judge_report.py [hunk 8]
```diff
@@ -25,6 +25,9 @@
     sys.exit(1)
 
 
+SKIP_SENTINEL = "Skipped:"
```
Comment (shafty023) [inline plugins/judges/skills/run-judges/scripts/validate_judge_report.py]: This sentinel is too broad for the skip contract. Combined with the empty-metrics exemption below, any `final_status=3` CaseScore containing `Skipped:` can validate as a skipped judge, even when it is not the fixed budget-overflow reason produced by `run-judges`. Please validate the exact skip reason, or define a structured skip taxonomy and add negative tests for unrelated `Skipped:` text.

File: plugins/code/.claude-plugin/plugin.json [hunk 9]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "code",
   "description": "Code and planning framework plugin",
-  "version": "1.11.20",
+  "version": "1.12.0",
```
Comment (shafty023) [inline plugins/code/.claude-plugin/plugin.json]: Current `origin/main` has the code plugin at `1.12.4`, but this branch sets it to `1.12.0`. If conflict resolution keeps this file, the release metadata moves backward. Please rebase or merge current `main` and bump forward from the current version according to the repo versioning rules.

File: .claude/code-review-findings.json [hunk 10]
```diff
@@ -0,0 +1,38 @@
+{"schema_version": 1, "pr_number": 69, "head_sha": "25482cee03f7c4520963bd6ca827300826ce4a73", "findings": [
```
Comment (shafty023) [inline .claude/code-review-findings.json]: This committed review artifact is for PR 69 and head `25482cee...`, not this PR. Please remove the `.claude/code-review-*` artifacts from the changeset, or regenerate them only if this repository intentionally tracks current PR review metadata.

File: plugins/judges/agents/context-manager-for-judges.md [hunk 11]
```diff
@@ -43,29 +44,65 @@ You will receive:
...
 ## Process
 
+### Step 0: Compute Token Budget
+
+Before any artifact collection, determine the token budget for this run:
+
+```bash
+ESTIMATED_OVERHEAD=98000
+
+if [[ -n "$CLOSEDLOOP_CONTEXT_LIMIT" ]]; then
```
Comment (shafty023) [inline plugins/judges/agents/context-manager-for-judges.md]: `CLOSEDLOOP_CONTEXT_LIMIT` is optional, but this snippet uses it directly in shell arithmetic. Invalid values such as non-numeric text can force a zero budget or abort the shell instead of preserving the default behavior. Please validate it as a positive decimal integer before arithmetic, warn/fall back when invalid, and mirror the same guard in the `run-judges` budget snippet.

File: plugins/judges/skills/run-judges/SKILL.md (hunk 6 above)
Comment (mikeangstadt) [reply plugins/judges/skills/run-judges/SKILL.md]: Fixed in 3531ea1 — the budget estimate no longer counts only the `judge-input.json` envelope. New `scripts/estimate_judge_budget.py` resolves every mapped artifact (`primary_artifact.path` + each `supporting_artifacts[].path`) relative to the workdir, sums them with the preambles plus an output reserve, applies the `/4` heuristic, and returns `skip_all_judges`. Sub-step B now calls it. Added a regression (`test_small_envelope_large_referenced_artifact_triggers_skip`) where the envelope is ~120 bytes but a referenced artifact is 50K tokens — it now correctly trips the skip guard.

File: plugins/judges/skills/run-judges/SKILL.md (hunk 7 above)
Comment (mikeangstadt) [reply plugins/judges/skills/run-judges/SKILL.md]: Fixed in 3531ea1 — Sub-step A now resolves the code-mode context file in mapper path order: it prefers the canonical `.closedloop-ai/context/code-context.json` and keeps root `code-context.json` only as a one-run legacy fallback. A normal code run can no longer read the root file first, miss `metadata.context_128k_mode`, and launch judges without the overflow guard.

File: plugins/judges/skills/run-judges/scripts/validate_judge_report.py (hunk 8 above)
Comment (mikeangstadt) [reply plugins/judges/skills/run-judges/scripts/validate_judge_report.py]: Fixed in 109d81a — replaced the broad `"Skipped:"` prefix sentinel with an exact-match taxonomy. `RECOGNIZED_SKIP_REASONS` (currently the single `BUDGET_SKIP_REASON = "Skipped: artifact exceeded context budget after compression"`) is matched exactly, so an unrelated `final_status=3` CaseScore that merely contains `Skipped:` no longer earns the empty-metrics exemption — it falls through to the normal metrics check. Added negative tests for arbitrary `Skipped:` text at both the top level and metric level, plus a validation-level test that such a report fails with "has no metrics".

File: plugins/code/.claude-plugin/plugin.json (hunk 9 above)
Comment (mikeangstadt) [reply plugins/code/.claude-plugin/plugin.json]: Fixed — rebased onto current `origin/main` (code plugin at 1.14.7) and bumped forward to **1.15.0** (MINOR, for the new `CLOSEDLOOP_CONTEXT_LIMIT` forwarding in run-loop.sh). The version no longer moves backward. Judges is at 1.8.0 (MINOR over main 1.7.1).

File: .claude/code-review-findings.json (hunk 10 above)
Comment (mikeangstadt) [reply .claude/code-review-findings.json]: Fixed in 500e711 — removed the `.claude/code-review-{findings.json,summary.md,threads.json}` artifacts (they were generated for PR 69 / head 25482cee, not this PR) and added a `.gitignore` guard (`.claude/code-review-*`) so per-PR review metadata is no longer tracked.

File: plugins/judges/agents/context-manager-for-judges.md (hunk 11 above)
Comment (mikeangstadt) [reply plugins/judges/agents/context-manager-for-judges.md]: Fixed in 3531ea1 — `CLOSEDLOOP_CONTEXT_LIMIT` is now validated as a positive decimal integer (`^[1-9][0-9]*$`) before any shell arithmetic, in both the run-judges Step 0.5 snippet and `context-manager-for-judges` Step 0. Non-numeric/empty/zero/negative values now warn and fall back to the default 200K-window behavior instead of forcing a zero budget or aborting the shell under `set -e`.
